core/src/backend/video/software.rs

```
@@ -198,3 +202,194 @@ mod h263 {
+        pub fn new(with_alpha: bool, bounds: (u16, u16)) -> Self {
+            // Unfortunately, `init()` cannot be called on the decoder
+            // just yet, because `bounds` is only the declared size of
+            // the video, to which it will be cropped.
+            // This can be (rarely) even much smaller than the actual
+            // encoded size of the frames.
+            // `VP56Decoder::init()` needs the full encoded frame size,
+            // so it can allocate its internal buffers accordingly.
+            // The encoded frame size will be parsed from the header of
+            // the first encoded frame passed to `Self::decode_frame()`.
+
+            Self {
+                with_alpha,
+                bounds,
+                decoder: VP56Decoder::new(6, with_alpha, true),
+                support: NADecoderSupport::new(),
+                bitreader: VP6BR::new(),
+                init_called: false,
+            }
+        }
+    }
+
+    impl VideoDecoder for Vp6Decoder {
+        fn preload_frame(
+            &mut self,
+            encoded_frame: EncodedFrame<'_>,
+        ) -> Result<FrameDependency, Error> {
+            // Luckily the very first bit of the encoded frames is exactly this flag,
+            // so we don't have to bother asking any "proper" decoder or parser.
+            Ok(
+                if !encoded_frame.data.is_empty() && (encoded_frame.data[0] & 0b_1000_0000) == 0 {
+                    FrameDependency::None
+                } else {
+                    FrameDependency::Past
+                },
+            )
+        }
+
+        fn decode_frame(&mut self, encoded_frame: EncodedFrame<'_>) -> Result<DecodedFrame, Error> {
+            // If this is the first frame, the decoder needs to be initialized.
+
+            if !self.init_called {
+                let mut bool_coder = BoolCoder::new(encoded_frame.data).unwrap();
+                let header = self.bitreader.parse_header(&mut bool_coder).unwrap();
```

Can we return the error here instead of `unwrap`?

---

I changed the `unwrap()`s to `.map_err(...)?` patterns.